src/main/java/domain/AbstractCatchOnMovePiece.java

```diff
@@ -0,0 +1,41 @@
+package domain;
+
+import static domain.PieceMoveResult.CATCH;
+import static domain.PieceMoveResult.FAILURE;
+import static domain.PieceMoveResult.SUCCESS;
+
+import java.util.Optional;
+
+abstract class AbstractCatchOnMovePiece extends AbstractPiece {
+    AbstractCatchOnMovePiece(Position position, Team team) {
+        super(position, team);
+    }
+
+    @Override
+    public final PieceMoveResult tryMove(Position targetPosition, PiecesOnChessBoard piecesOnChessBoard) {
+        Optional<PieceMoveResult> pieceMoveResult = tryMoveAssumeAloneAndCheckRoute(targetPosition, piecesOnChessBoard);
+        if (pieceMoveResult.isPresent()) {
+            return pieceMoveResult.get();
+        }
+        if (isMyTeam(targetPosition, piecesOnChessBoard)) {
+            return FAILURE;
+        }
+        if (isOtherTeam(targetPosition, piecesOnChessBoard)) {
+            return CATCH;
+        }
+        return SUCCESS;
+    }
+
+    protected abstract Optional<PieceMoveResult> tryMoveAssumeAloneAndCheckRoute(Position targetPosition,
+                                                                                 PiecesOnChessBoard piecesOnChessBoard);
+
+    private boolean isMyTeam(Position targetPosition, PiecesOnChessBoard piecesOnChessBoard) {
+        Optional<Team> targetTeam = piecesOnChessBoard.whichTeam(targetPosition);
+        return targetTeam.isPresent() && targetTeam.get().equals(getTeam());
```

Comment:
```suggestion
        return piecesOnChessBoard.whichTeam(targetPosition)
    .filter(team -> team.equals(getTeam()))
    .isPresent();
```
이렇게 써볼수있을거같아요 다른부분들도요